### ISSUE TITLE
Not possible to build dynamic annotations needed for Ambassodor

### ISSUE THREAD

**Issue by aberres** [kubernetes-sigs/kustomize#964]
When using [Ambassador](https://www.getambassador.io/) to dynamically route traffic it is necessary to add an annotation to every service.
Especially when running in a CI setup where review branches are deployed this annotation is completely dynamic but can be generated when knowing the service name and the namespace it is deployed into..

An annotation could look like this:
```yaml
metadata:
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind:  Mapping
      name: name_of_the_service
      prefix: /
      service: name_of_the_service.name_of_the_namespace_we_deploy_into
      host: a_configured_host
```

Right now we have this implemented with Helm. I am investigating right now how and if the workflow can be implemented with Kustomize.

As Kustomize is (nearly) template free I have the feeling that this is not easily to accomplish.

I was looking into transformer configurations but especially as the name of the referenced service is supposed to change with every release/branch I could not really wrap my head around it.

This is how the config would look like in Helm:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: {{ template "chart.fullname" . }}
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind:  Mapping
      name:  {{ template "chart.fullname" . }}
      prefix: /
      service: {{ template "chart.fullname" . }}.{{ .Release.Namespace }}
      host: {{ .Values.url }}
spec:
  type: ClusterIP
  ports:
    - port: 80
      targetPort: http
      protocol: TCP
      name: http
```

Is there any way/example how to achieve the same with Kustomize?

**Comment by jbrette** [comment]
Also I'm not convinced this is the best way to do it,
there is a solution using variables:

First create a dummy Values CRD or a ConfigMap
```yaml
apiVersion: v1
kind: Values
metadata:
  name: file1
spec:
  chart:
     fullname: name_of_the_service
  Release:
    Namespace: name_of_the_namespace_we_deploy_into
  url: a_configured_host
```
then create your service
```yaml
apiVersion: v1
kind: Service
metadata:
  name: $(Values.file1.spec.chart.fullname)
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind:  Mapping
      name:  $(Values.file1.spec.chart.fullname)
      prefix: /
      service: $(Values.file1.spec.chart.fullname).$(Values.file1.spec.Release.Namespace)
      host: $(Values.file1.spec.url)
spec:
  type: ClusterIP
  ports:
    - port: 80
      targetPort: http
      protocol: TCP
      name: http
```

declare variables in kustomization.yaml
```yaml
vars:
- name: Values.file1.spec.chart.fullname
  objref:
    apiVersion: v1
    kind: Values
    name: file1
  fieldref:
    fieldpath: spec.chart.fullname
...
```

```bash
kustomize build .
````

```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind:  Mapping
      name:  name_of_the_service
      prefix: /
      service: name_of_the_service.name_of_the_namespace_we_deploy_into
      host: a_configured_host
  name: name_of_the_service
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: http
  type: ClusterIP
```

**Comment by vsimon** [comment]
I just hit this same issue. I had originally tried defining a var and referencing it in the annotation but quickly noticed this did not work. For example:

in kustomization.yaml
```
...
vars:
- name: APP_SERVICE_NAME
  objref:
    kind: Service
    name: xxxx
    apiVersion: v1
```

in service.yaml
```
...
metadata:
  name: xxxxx
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind: Mapping
      name: $(APP_SERVICE_NAME)
      service: $(APP_SERVICE_NAME)
```

I'll try the way @jbrette suggested, but I suspect others will attempt the "obvious" way first and fail.

**Comment by dbarrosop** [comment]
I just tried this and found various problems:

1. I seem to be unable to reference annotations with slashes like the one in the example of this issue. The slash in the annotation name is processed and turned into a subkey.
2. If I ignore that issue and try to apply it with `kubectl -k` I get the error: `error: unable to recognize "build/cloud/kubernetes/overlays/beta": no matches for kind "Values" in version "v1"`
3. I found somewhere else that the API version is `apiVersion: kustomize.config.k8s.io/v1` but if I use that version jsonPatch seem to be unable to target it, I suspect due to the same issue as in (1)
